packages/yew/src/functional/append_vec.rs

```diff
@@ -0,0 +1,139 @@
+use std::cell::{Cell, UnsafeCell};
+
+pub struct AppendOnlyList<T: ?Sized> {
+    state: UnsafeCell<Vec<Box<T>>>,
+    counter: Cell<usize>,
+    #[cfg(debug_assertions)]
+    total_counter: Option<usize>,
+}
+
+#[cfg(test)]
+#[test]
+fn assert_state_list_not_sync() {
+    struct Invalid;
+    trait AmbiguousIfImpl<A> {
+        fn some_item() {}
+    }
+
+    impl<T: ?Sized> AmbiguousIfImpl<()> for T {}
+    impl<T: ?Sized + Sync> AmbiguousIfImpl<Invalid> for T {}
+    // This will fail to compile (ambiguous impl) if T: Sync ==> StateList<T>: Sync
+    // StateList<T> should never be sync!
+    fn _assert_generically<T: Sync>() {
+        let _ = <AppendOnlyList<T> as AmbiguousIfImpl<_>>::some_item;
+    }
+}
+
+impl<T: ?Sized> AppendOnlyList<T> {
+    pub fn new() -> Self {
+        Self {
+            state: UnsafeCell::default(),
+            counter: Cell::new(0),
+            #[cfg(debug_assertions)]
+            total_counter: None,
+        }
+    }
+
+    #[allow(clippy::mut_from_ref)]
+    pub fn next_state(&self, initializer: impl FnOnce() -> Box<T>) -> &mut T {
+        // SAFETY: Because we are not Sync, we don't need to atomically increase here
+        let position = self.counter.get();
+        self.counter.set(position + 1);
+        // SAFETY: we drop this ref before returning.
+        //  Because we are not Sync, no other thread can have concurrent access to this.
+        //  Because we assert that we don't re-enter in the initializer, this is legal
+        //   to access even afterwards.
+        let state = unsafe { &mut *self.state.get() };
+        assert!(position <= state.len(), "Detected re-entrant initializer()");
+        if position == state.len() {
+            state.push(initializer());
+        }
+        state[position].as_mut()
```

Comment:
This could use a comment explaining that this is safe WRT relocation because the vec items are boxed.

Btw, this generally sounds like a great use case for a linked list where we wouldn't have to worry about relocation.

Reply:
It's not as easy to use a `LinkedList` here. The `position` should be a O(1) lookup into the list for the current position. It's not always the back of the list. Storing some kind of `Cursor` borrowing the list seems even more unsafe. Using a `position` with a `LinkedList` will degrade performance on subsequent renders where the List is already established and is only traversed.

Reply:
You're right, I totally failed to consider the overhead  of random access for linked lists.